pkg/apis/configuration/validation/validation.go

```diff
@@ -707,6 +737,124 @@ func validateVariable(nVar string, validVars map[string]bool, fieldPath *field.P
 		msg := fmt.Sprintf("'%v' contains an invalid NGINX variable. Accepted variables are: %v", nVar, mapToPrettyString(validVars))
 		allErrs = append(allErrs, field.Invalid(fieldPath, nVar, msg))
 	}
+	return allErrs
+}
+
+func isValidSpecialVariableHeader(header string) []string {
+	errMsgs := validation.IsHTTPHeaderName(strings.Replace(header, "_", "-", -1))
+	if len(errMsgs) >= 1 || strings.Contains(header, "-") {
+		return []string{"a valid HTTP header must consist of alphanumeric characters or '_'"}
+	}
+	return nil
+}
+
+func validateSpecialVariable(nVar string, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	value := strings.SplitN(nVar, "_", 2)
+
+	switch value[0] {
+	case "arg":
+		for _, msg := range isArgumentName(value[1]) {
+			allErrs = append(allErrs, field.Invalid(fieldPath, nVar, msg))
+		}
+	case "http":
+		for _, msg := range isValidSpecialVariableHeader(value[1]) {
+			allErrs = append(allErrs, field.Invalid(fieldPath, nVar, msg))
+		}
+	case "cookie":
+		for _, msg := range isCookieName(value[1]) {
+			allErrs = append(allErrs, field.Invalid(fieldPath, nVar, msg))
+		}
+	}
+
+	return allErrs
+}
+
+func validateActionReturnCode(code int, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if (code >= 200 && code <= 299) || (code >= 400 && code <= 599) {
+		return allErrs
+	}
+
+	msg := fmt.Sprintf("must be a valid status code either 2XX, 4XX or 5XX, for example, 200 or 402.")
+	return append(allErrs, field.Invalid(fieldPath, code, msg))
+}
+
+func validateActionReturn(r *v1.ActionReturn, fieldPath *field.Path) field.ErrorList {
```

Comment:
this function isn't tested at all. could you add a unit test or extend a test like `TestValidateAction` ?